cmd/registry/controller/controller.go

```diff
@@ -144,107 +136,218 @@ func generateActions(
+		if visited[parent.GetResourceName().String()] {
+			continue
+		} else {
+			// Since the GeneratedResource is non-existent here,
+			// we will have to derive the exact name of the target resource
+			targetResourceName, err := resourceNameFromParent(resourcePattern, parent.GetResourceName().String())
 			if err != nil {
-				return nil, fmt.Errorf("cannot match resource with dependency. Error: %s", err.Error())
+				log.Errorf(ctx, "Cannot generate target resourceName to be created. Error: %s", err)
+				continue
 			}
 
-			if maxUpdateTime, ok := dMap[group]; ok {
-				// Take action if dependency timestamp is later than resource timestamp
-				if maxUpdateTime.After(resourceTime) {
-					takeAction = true
-				}
-				visited[group] = true
-			} else {
-				// For a given resource, each of it's defined dependency group should be present.
-				// If any one of the dependency groups is missing, avoid calculating any action for the resource
-				takeAction = false
-				break
-			}
-		}
+			a, err := evaluateAction(
+				targetResourceName,
+				time.Time{},
+				dependencyMaps,
+				generatedResource,
+				true,
+			)
 
-		if takeAction {
-			cmd, err := generateCommand(generatedResource.Action, resource.GetName())
 			if err != nil {
-				return nil, err
+				log.Errorf(ctx, "%s", err)
 			}
-			action := &Action{
-				Command:           cmd,
-				GeneratedResource: resource.GetName(),
-				RequiresReceipt:   generatedResource.Receipt,
+
+			if a != nil {
+				actions = append(actions, a)
 			}
-			actions = append(actions, action)
+
 		}
 	}
 
-	// Check patterns where resources do not exist in the registry. Here new resources will be generated
-	// for the dependencies which were not visited in the above loop.
-	// Iterate over first dependency source and evaluate that against remaining dependencies
-	if len(dependencyMaps) > 0 {
-		dMap0 := dependencyMaps[0]
-		for groupKey := range dMap0 {
-			takeAction := true
-			if _, ok := visited[groupKey]; !ok {
-				for _, dMap := range dependencyMaps {
-					// in a particular group resources for each dependencyPattern must exist
-					// for the controller to generate new targeted resource.
-					// Example:
-					// generated_resource: apis/-/artifacts/summary
-					// - dependencies:
-					//   - pattern: $resource.api/artifacts/complexity
-					//   - pattern: $resource.api/artifacts/vocabulary
-					// for a particular api group, "petstore"
-					// both apis/petstore/artifacts/complexity and apis/petstore/artifacts/vocabulary should be present
-					_, dependencyExists := dMap[groupKey]
-					if !dependencyExists {
-						takeAction = false
-						break
-					}
-				}
-			} else {
-				takeAction = false
-			}
+	return actions, nil
+}
+
+func evaluateAction(
+	targetResourceName ResourceName,
+	targetResourceTime time.Time,
+	dependencyMaps []map[string]time.Time,
+	generatedResource *rpc.GeneratedResource,
+	createMode bool) (*Action, error) {
```

Comment:
This seems like it should be split into separate functions for each supported "mode" because 1) it serves a different purpose in each mode, and 2) callers need to provide an arbitrary `targetResourceTime` argument to be ignored in create mode.

At first glance, it seems like create mode is:
* Get the entity key for each dependency.
* Validate that each dependency is in the dependency maps
* Generate an action for the resource.

and update mode is:
* Get the entity key for each dependency.
* Validate that each dependency is in the dependency maps.
* Generate an action for the resource if and only if it's stale.

These seem like they can be split into separate functions that are individually much simpler.